salt/modules/file.py

```diff
@@ -1421,58 +1421,59 @@ def replace(path,
-                else:
-                    result, nrepl = re.subn(cpattern, repl, r_data, count)
+            else:
+                # size of file in /proc is 0, but contains data
+                r_data =  "".join(r_file)
+            if search_only:
+                # Just search; bail as early as a match is found
+                if re.search(cpattern, r_data):
+                    return True  # `with` block handles file closure
+            else:
+                result, nrepl = re.subn(cpattern, repl, r_data, count)
 
-                    # found anything? (even if no change)
-                    if nrepl > 0:
+                # found anything? (even if no change)
+                if nrepl > 0:
+                    found = True
+                    # Identity check the potential change
+                    has_changes = True if pattern != repl else has_changes
+
+                if prepend_if_not_found or append_if_not_found:
+                    # Search for content, to avoid pre/appending the
+                    # content if it was pre/appended in a previous run.
+                    if re.search('^{0}$'.format(re.escape(content)),
+                                 r_data,
+                                 flags=flags_num):
+                        # Content was found, so set found.
                         found = True
-                        # Identity check the potential change
-                        has_changes = True if pattern != repl else has_changes
-
-                    if prepend_if_not_found or append_if_not_found:
-                        # Search for content, to avoid pre/appending the
-                        # content if it was pre/appended in a previous run.
-                        if re.search('^{0}$'.format(re.escape(content)),
-                                     r_data,
-                                     flags=flags_num):
-                            # Content was found, so set found.
-                            found = True
-
-                    # Keep track of show_changes here, in case the file isn't
-                    # modified
-                    if show_changes or append_if_not_found or \
-                       prepend_if_not_found:
-                        orig_file = r_data.read(filesize).splitlines(True)
-                        new_file = result.splitlines(True)
 
-        except (OSError, IOError) as exc:
-            raise CommandExecutionError(
-                "Unable to open file '{0}'. "
-                "Exception: {1}".format(path, exc)
-                )
-        finally:
-            if r_data and isinstance(r_data, mmap.mmap):
-                r_data.close()
+                # Keep track of show_changes here, in case the file isn't
+                # modified
+                if show_changes or append_if_not_found or \
+                   prepend_if_not_found:
+                    orig_file = r_data.read(filesize).splitlines(True)
```

Comment:
It feels most robust to me just to try `r_data = r_file` instead of the join. But if that doesn't work for whatever reason, maybe something like this, here?

```
orig_file = r_data.read(filesize).splitlines(True) if hasattr(r_data, 'read') \
    else r_data.splitlines(True)
```